src/python/pants/util/rwbuf.py

```diff
@@ -6,6 +6,7 @@
                         unicode_literals, with_statement)
 
 import threading
+from builtins import object, str
```

Comment:
@benjyw can you look over this file please? 

This override means that in line 39, during call to `_RWBuf.write()`, it will call `_RWBuf.do_write()` with `unicode`, rather than `bytes`, which might be a problem. However, `do_write()` simply raises `NotImplementedError`, so I don't know if this actually matters?

Reply:
If I understand correctly, `from builtins import str` is a no-op in py3 but renames py2's unicode type to `str` in py2?  I think this definitely will have consequences. A low-level IO method should always operate on raw bytes (`str` in py2, `bytes` in py3), never on encoded strings.  

However this file only explicitly uses `str` in one line, and that's hacky:

`self.do_write(str(s))`

If `s` is already binary data, no problem, but if it's text, then in python2 that `str(s)` would attempt to encode it using the default encoding, which is not guaranteed to work. 

So really write() should check if the argument is bytes, and throw an exception of not. That way we'll find out eagerly if any call sites do the wrong thing.

Does that make sense?

Reply:
I imagine something like `from builtins import bytes` and then in `write()` check for `isinstance(s, bytes)`.

Reply:
Yes, exactly. `from builtins import str` modifies Python2's `str()` function so that it gets a unicode, rather than bytes, representation.

Agreed with that change. I'll break it out into a separate PR since it's a substantial semantic change + we have no explicit tests for this file.

Reply:
Actually what I'm proposing involves no semantic change. This commit, as stands, would be a major semantic change, as it would cast arbitrary bytes into text, which might explode. So you can't commit this as-is.